feat(AIP-121): adding guidance around consistency

toumorokoshi opened the pull request:
Google clients have implicit requirements around the consistency today, but it is not explicitly stated in the AIPs.

Clarifying the guidance around consistency requirements.

toumorokoshi committed:
feat(AIP-121): adding guidance around consistency

Google clients have implicit requirements around the
consistency today, but it is not explicitly stated in the AIPs.

Clarifying the guidance around consistency requirements.

bgrant0607 commented on aip/general/0121.md:
```diff
@@ -94,6 +94,33 @@ the standard methods. Custom methods offer the same design freedom as
 traditional RPC APIs, which can be used to implement common programming
 patterns, such as database transactions, import and export, or data analysis.
 
+### Consistency
```
"Read-after-write consistency"

Otherwise "consistency" is a very overloaded term.

bgrant0607 replied:
Or maybe "strong consistency".

Do we care about write-after-write consistency also?

toumorokoshi replied:
Definition-wise I was hoping to leverage the term from the CAP theorem: https://en.wikipedia.org/wiki/CAP_theorem: "Every read receives the most recent write or an error." Although I imagine we have to qualify this a bit better.

Or I could title the section "Consistency Guarantees", since that generally implies distributed systems?

write-after-write... I didn't think it really mattered, since one can always validate whether their intended state has been applied and re-read again. 

For those who do want to have guarantee around their resource state's write, I figured AIP-154's etag might be helpful?

bgrant0607 commented on aip/general/0121.md:
```diff
@@ -94,6 +94,33 @@ the standard methods. Custom methods offer the same design freedom as
+### Consistency
+
+For methods that operate on a resource, the completion of those operations
+(either successful or with an error) **must** mean that the state of the
```
Do we need to mention LRO here, or is that understood?

toumorokoshi replied:
I feel like it's understood, since "long-running operation" is a variation of "operation", after all.

But I see the value in clarifying that it's both.

bgrant0607 commented on aip/general/0121.md:
```diff
@@ -94,6 +94,33 @@ the standard methods. Custom methods offer the same design freedom as
+- A get request for a resource immediately after the successful completion of a
+  create method **must** return the resource.
+- A get request for a resource immediately following an update **must** return
+  a resource with the final values from the update request.
```
What about delete?

toumorokoshi replied:
ack. These were just examples, but I suppose it doesn't hurt to add a delete.

bgrant0607 commented on aip/general/0134.md:
```diff
@@ -60,6 +60,12 @@ rpc UpdateBook(UpdateBookRequest) returns (Book) {
   - All remaining fields **should** map to URI query parameters.
 - There **should** be exactly one `google.api.method_signature` annotation,
   with a value of `"{resource},update_mask"`.
+- The operation should be [consistent][]: the completion of an update operation
```
"strongly consistent"

bgrant0607 replied:
"read-after-write consistent" sounds odd to me. Probably would need to change the verb if we go with that.

toumorokoshi replied:
strongly consistent works here, will update.

bgrant0607 commented on aip/general/0135.md:
```diff
@@ -53,6 +53,11 @@ rpc DeleteBook(DeleteBookRequest) returns (google.protobuf.Empty) {
 - There **should** be exactly one `google.api.method_signature` annotation,
   with a value of `"name"`. If an etag or force field are used, they **may** be
   included in the signature.
+- The operation should be [consistent][]: the completion of a delete operation
```
Same here.

toumorokoshi committed:
addressing feedback

- normalizing wording to "strong consistency", to help clarify that
  this is referring to distributed system consistency.
- clarifying LRO scope for operation completion.

toumorokoshi committed:
Merge branch 'master' into toum/add-consistency

bgrant0607 reviewed (approved):
LGTM. Thanks for the changes. This is clearer.

toumorokoshi committed:
Merge branch 'master' into toum/add-consistency

toumorokoshi committed:
Merge branch 'master' into toum/add-consistency

toumorokoshi committed:
address comments

adding caveat around management plane for strong consistency, to allow
some data-plane use cases of standard methods like eventual consistency
due to large volumes of calls.

alin04 commented on aip/general/0121.md:
```diff
@@ -94,6 +94,37 @@ the standard methods. Custom methods offer the same design freedom as
 
+### Strong Consistency
+
+For methods that operate on the [managment plane][], the completion of those
```
```suggestion
For methods that operate on the [management plane][], the completion of those
```

toumorokoshi replied:
applied.

alin04 commented on aip/general/0121.md:
```diff
@@ -94,6 +94,37 @@ the standard methods. Custom methods offer the same design freedom as
+- Following a successful create, a get request for a resource **must**
+  return the resource.
+- Following a successful update, a get request for a resource **must**
+  return the finale values from the update request.
```
```suggestion
  return the final values from the update request.
```

toumorokoshi replied:
applied.

alin04 commented on aip/general/0121.md:
```diff
@@ -94,6 +94,37 @@ the standard methods. Custom methods offer the same design freedom as
+mean that the state of the resource's existence and all user-settable values
+have reached a steady-state.
+
+[output only][] values **should** also have reached a steady-state, except
```
What does this mean?

toumorokoshi replied:
basically that output-only values should be stable if they aren't related to resource state. E.g. they're not jittering around in values. This is for the same reason the operation itself should be consistent: the user should expect resource creation marks some event where the resource is more or less stable, since declarative clients have trouble reconciling transient state, and users will be surprised if the resource is constantly changing.

Of course being related to resource state is a little vague, since the only definition of state we have is AIP-216.

alin04 replied:
Like server generated IDs? create_time ?

toumorokoshi replied:
those are good examples of things that are steady state. AIP-216's state field is an example of what shouldn't be.

Would adding examples help? Or do you think this statement is too ambiguous?

toumorokoshi committed:
addressing comments

alin04 commented on aip/general/0121.md:
```diff
@@ -94,6 +94,37 @@ the standard methods. Custom methods offer the same design freedom as
+
+Examples include:
+
+- Following a successful create, a get request for a resource **must**
```
Unless there's a race condition whereby it got deleted immediately. Same below for update.

If two people are updating, (unless the service handles stale state), then last one will win.

toumorokoshi replied:
good call! I updated the working to state that it has to be the latest mutation on the resource (should clarify that last operation wins). WDYT?

alin04 commented on aip/general/0121.md:
```diff
@@ -94,6 +94,37 @@ the standard methods. Custom methods offer the same design freedom as
+mean that the state of the resource's existence and all user-settable values
+have reached a steady-state.
+
+[output only][] values **should** also have reached a steady-state, except
```
```suggestion
[output only][] values unrelated to the resource [state][] **should** also have reached a steady-state.
```

toumorokoshi replied:
applied.

toumorokoshi committed:
addressing feedback

toumorokoshi committed:
Merge branch 'master' into toum/add-consistency
